src/include/sof/lib/dma.h

```diff
@@ -281,141 +282,253 @@ void dma_put(struct dma *dma);
+	platform_shared_commit(channel, sizeof(*channel));
+
+	return ret;
 }
 
 static inline int dma_status(struct dma_chan_data *channel,
 			     struct dma_chan_status *status, uint8_t direction)
 {
-	return channel->dma->ops->status(channel, status, direction);
+	int ret = channel->dma->ops->status(channel, status, direction);
+
+	platform_shared_commit(channel->dma, sizeof(*channel->dma));
+	platform_shared_commit(channel, sizeof(*channel));
+
+	return ret;
 }
 
 static inline int dma_set_config(struct dma_chan_data *channel,
 				 struct dma_sg_config *config)
 {
-	return channel->dma->ops->set_config(channel, config);
+	int ret = channel->dma->ops->set_config(channel, config);
+
+	platform_shared_commit(channel->dma, sizeof(*channel->dma));
+	platform_shared_commit(channel, sizeof(*channel));
+
+	return ret;
 }
 
 static inline int dma_pm_context_restore(struct dma *dma)
 {
-	return dma->ops->pm_context_restore(dma);
+	int ret = dma->ops->pm_context_restore(dma);
+
+	platform_shared_commit(dma, sizeof(*dma));
+
+	return ret;
 }
 
 static inline int dma_pm_context_store(struct dma *dma)
 {
-	return dma->ops->pm_context_store(dma);
+	int ret = dma->ops->pm_context_store(dma);
+
+	platform_shared_commit(dma, sizeof(*dma));
+
+	return ret;
 }
 
 static inline int dma_probe(struct dma *dma)
 {
-	return dma->ops->probe(dma);
+	int ret = dma->ops->probe(dma);
+
+	platform_shared_commit(dma->chan, sizeof(*dma->chan) *
+			       dma->plat_data.channels);
+	platform_shared_commit(dma, sizeof(*dma));
+
+	return ret;
 }
 
 static inline int dma_remove(struct dma *dma)
 {
-	return dma->ops->remove(dma);
+	int ret = dma->ops->remove(dma);
+
+	platform_shared_commit(dma, sizeof(*dma));
+
+	return ret;
 }
 
 static inline int dma_get_data_size(struct dma_chan_data *channel,
 				    uint32_t *avail, uint32_t *free)
 {
-	return channel->dma->ops->get_data_size(channel, avail, free);
+	int ret = channel->dma->ops->get_data_size(channel, avail, free);
+
+	platform_shared_commit(channel->dma, sizeof(*channel->dma));
+	platform_shared_commit(channel, sizeof(*channel));
+
+	return ret;
 }
 
 static inline int dma_get_attribute(struct dma *dma, uint32_t type,
 				    uint32_t *value)
 {
-	return dma->ops->get_attribute(dma, type, value);
+	int ret = dma->ops->get_attribute(dma, type, value);
+
+	platform_shared_commit(dma, sizeof(*dma));
+
+	return ret;
 }
 
 static inline int dma_interrupt(struct dma_chan_data *channel,
 				enum dma_irq_cmd cmd)
 {
-	return channel->dma->ops->interrupt(channel, cmd);
+	int ret = channel->dma->ops->interrupt(channel, cmd);
+
+	platform_shared_commit(channel->dma, sizeof(*channel->dma));
+	platform_shared_commit(channel, sizeof(*channel));
+
+	return ret;
 }
 
 /* DMA hardware register operations */
 static inline uint32_t dma_reg_read(struct dma *dma, uint32_t reg)
 {
-	return io_reg_read(dma_base(dma) + reg);
+	uint32_t val;
+
+	val = io_reg_read(dma_base(dma) + reg);
+
+	platform_shared_commit(dma, sizeof(*dma));
```

Comment:
Same comment here, I don't think commit is needed if the struct isn't modified but only read?

Reply:
@paulstelian97 It's a little bit more tricky. In order to get things synchronized we need to have data invalidated before read. Reading member of dma retrieves data to cache, so without commit we won't see any potential updates which will happen between this and next read (commit should do writeback and invalidate). The only other possibility to keep it synced would be to call platform_shared_get() first, which would do invalidation in such case, but that would require two calls instead of one in the places, where we actually change data ( get + commit).

Reply:
Ah I see, so it's to ensure future reads get fresh data (someone else can modify dma_base(dma)?)

I don't see the value of doing commits on read only data (didn't you later make the dma struct const anyway?)

Reply:
(the commit won't affect the _actual_ data being read by these I/O operations)

Reply:
@paulstelian97 This is actually something I mentioned in the PR description and what I'll try to do soon: Some of the structures like dai and dma should be divided into shared and not shared parts, since most of the data is constant.
Right now constant data lies next to non constant, so reading dma will get the whole cache line.

Reply:
Alright, do respond to the comment about spinlocks in uncached areas (do they work fine on Xtensa?) so I can approve this.